File: Part2-Regression/Section 5 - Multiple Linear Regression/multiple_linear_regression.R
# Multiple Linear Regression

# Importing the dataset
dataset = read.csv('50_Startups.csv')

# Encoding categorical data
dataset$State = factor(dataset$State,
                         levels = c('New York', 'California', 'Florida'),
                         labels = c(1, 2, 3))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fittinh Multiple Linear Regression to the Training set
regressor = lm(formula = Profit ~ ., data = training_set)

# Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)

# Building the optimal model using Backward Elimination
SL = 0.05
Pv = 100
x = dataset[, c(1, 2, 3, 4, 5)]
regressor = lm(formula = Profit ~ ., data = x)
while (Pv > SL) {
  numVars = length(x)
  regressor = lm(formula = Profit ~ ., data = x)
  maxPv = max(coef(summary(regressor))[c(2: numVars), "Pr(>|t|)"])
  maxPvIndex = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxPv)
  x = x[, -maxPvIndex]
  Pv = maxPv
}

summary(regressor)

File: Part2-Regression/Section 5 - Multiple Linear Regression/multiple_linear_regression.py
# Multiple Linear Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Encoding the Independent Variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Fitting Multiple Linear Regression to The Training set
from sklearn.linear_model import  LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

# use this to show plot in new window
# %matplotlib qt

# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1)

SL = 0.05
Pv = 1

X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
while (Pv > SL):

    regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
    regressor_OLS.summary()
    xLen = X_opt[0].__len__()
    Pv = max(regressor_OLS.pvalues).astype(float)
    print(Pv)
    if (Pv > SL):
        for j in range(0, xLen):
            if (regressor_OLS.pvalues[j].astype(float) == Pv):
                X_opt = np.delete(X_opt, j, 1)

regressor_OLS.summary()
